### PR TITLE
fix(Octree): fix octree memleak

### PR DESCRIPTION
**Pull request by BrentHuang**
https://github.com/cdcseacave/openMVS/issues/896

### CODE REVIEW

File: libs/Common/Octree.inl [hunk 1]
```diff
@@ -167,6 +174,8 @@ void TOctree<ITEMARR_TYPE,TYPE,DIMS,DATA_TYPE>::_Insert(CELL_TYPE& cell, const P
 	// if this child cell needs to be divided further
 	if (bForceSplit || insertData.split(size, radius)) {
 		// init node and proceed recursively
+		cell.Release();
```
Comment (cdcseacave) [inline libs/Common/Octree.inl:177]: cell should be always empty, pls remove the release and add an assert instead:
```
ASSERT(cell.m_child == NULL);
```

File: libs/MVS/SceneRefine.cpp [hunk 2]
```diff
@@ -424,6 +424,7 @@ void MeshRefine::ListCameraFaces()
 			Mesh::FacesInserter inserter(arrCameraFaces[ID]);
 			octree.Traverse(frustum, inserter);
 		}
+        octree.Release();
```
Comment (cdcseacave) [inline libs/MVS/SceneRefine.cpp:427]: none of these explicit releases are needed, the memory is released automatically by the destructor

File: libs/Common/Octree.inl [hunk 3]
```diff
@@ -25,16 +25,23 @@ inline TOctree<ITEMARR_TYPE,TYPE,DIMS,DATA_TYPE>::CELL_TYPE::CELL_TYPE()
...
 } // destructor
 /*----------------------------------------------------------------*/
 
 
 template <typename ITEMARR_TYPE, typename TYPE, int DIMS, typename DATA_TYPE>
 inline void TOctree<ITEMARR_TYPE,TYPE,DIMS,DATA_TYPE>::CELL_TYPE::Release()
 {
-	delete[] m_child;
-	m_child = NULL;
+    if (m_child != NULL)
```
Comment (cdcseacave) [inline libs/Common/Octree.inl:40]: the if is not needed, it is done automatically inside delete

File: libs/Common/Octree.inl [hunk 4]
```diff
@@ -25,16 +25,23 @@ inline TOctree<ITEMARR_TYPE,TYPE,DIMS,DATA_TYPE>::CELL_TYPE::CELL_TYPE()
 template <typename ITEMARR_TYPE, typename TYPE, int DIMS, typename DATA_TYPE>
 inline TOctree<ITEMARR_TYPE,TYPE,DIMS,DATA_TYPE>::CELL_TYPE::~CELL_TYPE()
 {
-	delete[] m_child;
+    if (m_child != NULL)
```
Comment (cdcseacave) [inline libs/Common/Octree.inl:28]: if not needed, and setting it to null neither, as the object is destructed anyway, revert

**Review by cdcseacave — CHANGES_REQUESTED** [review]

File: libs/Common/Octree.inl (hunk 1 above)
Comment (BrentHuang) [reply libs/Common/Octree.inl:177]: Selection_002
Selection_003

File: libs/Common/Octree.inl (hunk 4 above)
Comment (BrentHuang) [reply libs/Common/Octree.inl:28]: Do different compilers behave the same way?

File: libs/MVS/SceneRefine.cpp (hunk 2 above)
Comment (BrentHuang) [reply libs/MVS/SceneRefine.cpp:427]: If you rely on destructors, it is recommended to make the Release method private

File: libs/Common/Octree.inl (hunk 1 above)
Comment (cdcseacave) [reply libs/Common/Octree.inl:177]: I see, so pls fix the double allocation instead of just adding a release

File: libs/MVS/SceneRefine.cpp (hunk 2 above)
Comment (cdcseacave) [reply libs/MVS/SceneRefine.cpp:427]: no, Release() is still useful to call manually when you want to reuse the object, but it is also called automatically at destruction

File: libs/Common/Octree.inl (hunk 4 above)
Comment (cdcseacave) [reply libs/Common/Octree.inl:28]: yes

File: libs/Common/Octree.inl (hunk 1 above)
Comment (BrentHuang) [reply libs/Common/Octree.inl:177]: valgrind reported a memory leak. I see that the code does have new twice, but I'm not sure why.

**Comment by cdcseacave** [comment]
thanks for spotting this issue, but this PR is not a valid fix, I fix it in bcad1eaadff463bbb7995d19f9eeae3e818a99c2

**Pull request closed without merging**